### PR TITLE
Support geo label position through REST vector tiles API

### LINKED ISSUE #86044: Add label features to vector tile search API response
### Description

There are [2 labeling issues in Kibana when using Vector tiles](https://github.com/elastic/kibana/issues/130675).
1) Multiple labels for MultiPolygon and MultiLineString geometries.
    Screen Shot 2022-04-20 at 9 32 26 AM
2) Multiple labels for features that span tiles. For example, hexagons that span tile boundaries.
    Screen Shot 2022-04-20 at 9 35 52 AM

The recommended solution is to provide features for labels, where the feature geometry is the label location.

To resolve these issues, I propose adding a new parameter, `withLabels`,  to `_mvt` API that when `true` would add label features to the response. Label features could be added in 2 ways
1) another layer, "label",  in the response. The "label" layer will contain Point feature for each hit and agg feature. For agg features, the Point feature will be the center of the hexagon. For hit features, the Point feature will be the centroid of the largest polygon for MultiPolygon and the centroid for the longest line for MultiLineString.
2) Include label feature in existing "hits" and "aggs" layers. Give label feature a property like "__EsIsLabelFeature" so filter expressions can be used exclude features in style rules.

Kibana is creating these centroid features today for GeoJSON sources. Kibana includes label features in GeoJSON feature collection and uses filters to exclude features in style rules (option 2 above) 

cc @jsanz @thomasneirynck

### PR DESCRIPTION
**Pull request by craigtaverner**
There is a need to provide sensibly calculated label positions for polygons and lines in Kibana maps (as requested in https://github.com/elastic/elasticsearch/issues/86044). A very convenient way to satisfy this need is through a runtime field that the rest API can make use of when labels are requested. This has the advantage of providing painless access to the label position as well.

Building onto https://github.com/elastic/elasticsearch/pull/86154, this PR adds support for the REST API to provide label positions to MVT queries, both for the HITS layer and the AGGS layer. To enable this feature, set `with_labels` to `true` as a query parameter to the vector tile search query.

Closes #86044

### CODE REVIEW

**Review by iverase — CHANGES_REQUESTED** [review]
I think we need to add a yaml test here: https://github.com/elastic/elasticsearch/blob/9a60aa8b64209740616266c2607b4600d90f3b6b/x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/vector-tile/10_basic.yml

apart from that I tried locally and it is looking great!

**Comment by iverase** [comment]
In addition, can we update the description of the PR because it is used to generate the release notes?

File: x-pack/plugin/vector-tile/src/main/java/org/elasticsearch/xpack/vectortile/rest/RestVectorTileAction.java [hunk 1]
```diff
@@ -274,6 +297,21 @@ private static VectorTile.Tile.Layer.Builder buildHitsLayer(SearchHit[] hits, Ve
...
+                if (labelField != null) {
+                    Object labelPosValue = labelField.getValue();
+                    GeoPoint labelPos = GeoUtils.parseGeoPoint(labelPosValue, true);
+                    byte[] labelPosFeature = featureFactory.point(labelPos.lon(), labelPos.lat());
+                    if (labelPosFeature != null && labelPosFeature.length != 0) {
+                        featureBuilder.clear();
+                        featureBuilder.mergeFrom(labelPosFeature);
+                        VectorTileUtils.addPropertyToFeature(featureBuilder, layerProps, ID_TAG, searchHit.getId());
+                        VectorTileUtils.addPropertyToFeature(featureBuilder, layerProps, LABEL_POSITION_TAG, true);
+                        hitsLayerBuilder.addFeatures(featureBuilder);
```
Comment (iverase) [inline x-pack/plugin/vector-tile/src/main/java/org/elasticsearch/xpack/vectortile/rest/RestVectorTileAction.java:311]: I think we agreed that we should add all the tags that we added to the feature that is not a tag. That brings the question to what happens if the user has defined an attribute with the same name as the label?

File: x-pack/plugin/vector-tile/src/main/java/org/elasticsearch/xpack/vectortile/rest/RestVectorTileAction.java [hunk 2]
```diff
@@ -309,6 +346,18 @@ private static VectorTile.Tile.Layer.Builder buildAggsLayer(
...
+                // Add label position as point
+                featureBuilder.clear();
+                GeoPoint labelPos = (GeoPoint) bucket.getKey();
+                byte[] labelPosFeature = featureFactory.point(labelPos.lon(), labelPos.lat());
+                if (labelPosFeature != null && labelPosFeature.length != 0) {
+                    featureBuilder.mergeFrom(labelPosFeature);
+                    VectorTileUtils.addPropertyToFeature(featureBuilder, layerProps, KEY_TAG, bucketKey);
+                    VectorTileUtils.addPropertyToFeature(featureBuilder, layerProps, LABEL_POSITION_TAG, true);
+                    aggLayerBuilder.addFeatures(featureBuilder);
+                }
```
Comment (iverase) [inline x-pack/plugin/vector-tile/src/main/java/org/elasticsearch/xpack/vectortile/rest/RestVectorTileAction.java:359]: I think we agreed that we should add all the tags that we added to the feature that is not a tag. That brings the question to what happens if the user has defined an attribute with the same name as the label?

**Review by iverase — CHANGES_REQUESTED** [review]

File: x-pack/plugin/vector-tile/src/main/java/org/elasticsearch/xpack/vectortile/rest/RestVectorTileAction.java (hunk 1 above)
Comment (craigtaverner) [reply x-pack/plugin/vector-tile/src/main/java/org/elasticsearch/xpack/vectortile/rest/RestVectorTileAction.java:309]: I think both concerns are considered in https://github.com/elastic/elasticsearch/pull/86458/commits/9fa8c0028b756e904beeeddc8a95c0cc4707fbe4

**Review by iverase — APPROVED** [review]
LGTM!

File: docs/reference/search/search-vector-tile-api.asciidoc [hunk 3]
```diff
@@ -353,6 +358,22 @@ If `true`, the exact number of hits is returned at the cost of some performance.
...
+If true, the hits and aggs layers will contain additional point features representing
+suggested label positions for the original features.
+* `Point` and `MultiPoint` features will have one of the points selected.
+* `Polygon` and `MultiPolygon` features will have a single point generated,
+either the centroid, if it is within the polygon, or another point within the
+polygon selected from the <<geoshape-indexing-approach,sorted triangle-tree>>.
+* `LineString` features will likewise provide a roughly central point selected
+from the <<geoshape-indexing-approach,triangle-tree>>.
+* The aggregation results will provide one central point for each aggregation bucket.
+
```
Comment (nreese) [inline docs/reference/search/search-vector-tile-api.asciidoc:373]: Use `+` instead of empty line so statement below is indented under with_labels option

Comment (iverase) [reply docs/reference/search/search-vector-tile-api.asciidoc:373]: I will add that the label features inherit all the attributes from its corresponding original feature.

Comment (craigtaverner) [reply docs/reference/search/search-vector-tile-api.asciidoc:373]: I had tried that before, but it indented too much (made it part of the last bullet point). However, now I tried again, this time with an extra newline before the `+` and voila! It indented the right amount!

Comment (craigtaverner) [reply docs/reference/search/search-vector-tile-api.asciidoc:373]: I fixed both your suggestions in https://github.com/elastic/elasticsearch/pull/86458/commits/1c69a4aec99dea1891b087c1ed711c730729a0ee

**Review by nreese — APPROVED** [review]
doc changes LGTM

**Pull request merged**